### ISSUE TITLE
docker4mac startup issues involving NO_PROXY

### ISSUE THREAD

**Issue by kastork** [eclipse-che/che#10213]
### Description

Che complains about NO_PROXY needing to be set, but every effort I've made to set a value seems to be ignored.

### Reproduction Steps

macos 10.13.5
docker4mac 18.03.1-ce-mac65 (24312) network in "bridge" mode.
There is no proxy running on macos and the system control panel for proxies has no selected items for local or remote proxies.

```
› docker run -ti -e NO_PROXY=localhost -v /var/run/docker.sock:/var/run/docker.sock -v ~/Documents/che-data1:/data eclipse/che:6.7.1 start
INFO: Proxy: HTTP_PROXY=docker.for.mac.http.internal:3128, HTTPS_PROXY=docker.for.mac.http.internal:3129, NO_PROXY=
WARN: Potential networking issue discovered!
WARN: We have identified that http and https proxies are set but no_proxy is not. This may cause fatal networking errors. Set no_proxy for your Docker daemon!
INFO: (che cli): 6.7.1 - using docker 18.03.1-ce / docker4mac
INFO: (che config): Generating che configuration...
INFO: (che config): Customizing docker-compose for running in a container
INFO: (che start): Preflight checks
         mem (1.5 GiB):           [OK]
         disk (100 MB):           [OK]
         port 8080 (http):        [AVAILABLE]
         conn (browser => ws):    [OK]
         conn (server => ws):     [OK]

INFO: (che start): Starting containers...
INFO: (che start): Services booting...
INFO: (che start): Server logs at "docker logs -f che"
INFO: (che start): Booted and reachable
INFO: (che start): Ver: 6.7.1
INFO: (che start): Use: http://localhost:8080
INFO: (che start): API: http://localhost:8080/swagger
```
I get the same results if I use `CHE_NO_PROXY=localhost`

As you can see, startup does complete.  However the warning about no_proxy is a good one, because starting a workspace fails on what appears to be an inability to pull an image.

I have also tried making adjustments to the macos system settings for proxies. But these do not appear to be honored by docker4mac.

**Diagnostics:** 

```
2018-06-27 12:00:59,940[nio-8080-exec-4]  [INFO ] [o.e.c.a.w.s.WorkspaceRuntimes 329]   - Starting workspace 'che/wksp-zi5a' with id 'workspaceiyw8vhpwlo4jd5sr' by user 'che'
2018-06-27 12:01:16,442[equestProcessor]  [ERROR] [l.ErrorInstallerLogEventLogger 42]   - Installer `org.eclipse.che.ws-agent` error from machine=`dev-machine` owner=`che` env=`default` workspace=`workspaceiyw8vhpwlo4jd5sr` text=`Debugger failed to attach: handshake failed - connection prematurally closed` time=`2018-06-27T12:01:16.427626807Z`
2018-06-27 12:04:19,286[aceSharedPool-0]  [INFO ] [o.e.c.a.w.s.WorkspaceRuntimes 390]   - Workspace 'che:wksp-zi5a' with id 'workspaceiyw8vhpwlo4jd5sr' start failed
```

I have seen other issues about `NO_PROXY` and docker4mac and/or che.

https://github.com/eclipse/che/issues/8368

https://github.com/docker/for-mac/issues/2715 (docker issue)

https://github.com/docker/for-mac/issues/2732 (docker issue that refers specifically to che)

None of which seem to provide resolution or effective workarounds.

The bottom line is that the warning issued by che is good, but there's no solution in messages or documentation as to how to communicate a value that che wants when trying to run it on docker4mac.  Both no help in saying how to set a value, and no help in determining what that value should be.

**Comment by eivantsov** [comment]
@kastork have you tried to export NO_PROXY so that your Docker daemon picks it up?

What's the output of `docker info`?

**Comment by kastork** [comment]
docker4mac picks up the proxy info from the macos System Preferences. If you manually put in proxy values there (remember, I don't actually have proxy values that I want to use for HTTP and HTTPS proxies, so they would be blank), you goof up the internal docker4mac HTTP and HTTPS proxies.

Here's my `docker info`

```
› docker info
Containers: 8
 Running: 0
 Paused: 0
 Stopped: 8
Images: 52
Server Version: 18.03.1-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host ipvlan macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 773c489c9c1b21a6d78b5c538cd395416ec50f88
runc version: 4fc53a81fb7c994640722ac585fa9ca548971871
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.87-linuxkit-aufs
Operating System: Docker for Mac
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 1.952GiB
Name: linuxkit-025000000001
ID: EUUU:QCFW:S6PG:BDH6:7OYG:K3EX:FNB7:C5S7:I242:62Z5:F7SC:ZDT4
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 24
 Goroutines: 42
 System Time: 2018-06-27T18:40:51.436455885Z
 EventsListeners: 2
HTTP Proxy: docker.for.mac.http.internal:3128
HTTPS Proxy: docker.for.mac.http.internal:3129
Registry: https://index.docker.io/v1/
Labels:
Experimental: true
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Comment by kastork** [comment]
Has there been any work on this?  Still seeing it in 6.10

**Comment by prcamp** [comment]
I'm running into the same issue trying to run che on mac, with a similar configuration as @kastork - super frustrating.

**Comment by kastork** [comment]
This may or may not be helpful.

gist.github.com/kastork/3d060691e9661064e894c2aa1893e0f4

**Comment by eivantsov** [comment]
@kastork Che picks up Docker daemon proxy settings. 

```
# Check to see if HTTP_PROXY, HTTPS_PROXY, and NO_PROXY is set within the Docker daemon.
  OUTPUT=$(docker info)
  HTTP_PROXY=$(grep "Http Proxy" <<< "$OUTPUT" || true) 
  if [ ! -z "$HTTP_PROXY" ]; then
    HTTP_PROXY=${HTTP_PROXY#"Http Proxy: "}
  else 
    HTTP_PROXY=""
  fi

  HTTPS_PROXY=$(grep "Https Proxy" <<< "$OUTPUT" || true) 
  if [ ! -z "$HTTPS_PROXY" ]; then
    HTTPS_PROXY=${HTTPS_PROXY#"Https Proxy: "}
  else
    HTTPS_PROXY=""
  fi

  NO_PROXY=$(grep "No Proxy" <<< "$OUTPUT" || true) 
  if [ ! -z "$NO_PROXY" ]; then
    NO_PROXY=${NO_PROXY#"No Proxy: "}
  else
    NO_PROXY=""
  fi

  if [[ ! ${HTTP_PROXY} = "" ]] ||
     [[ ! ${HTTPS_PROXY} = "" ]] ||
     [[ ! ${NO_PROXY} = "" ]]; then
     info "Proxy: HTTP_PROXY=${HTTP_PROXY}, HTTPS_PROXY=${HTTPS_PROXY}, NO_PROXY=${NO_PROXY}"
     if [[ ${NO_PROXY} = "" ]]; then
     warning "Potential networking issue discovered!"
     warning "We have identified that http and https proxies are set but no_proxy is not. This may cause fatal networking errors. Set no_proxy for your Docker daemon!"
     fi
  fi

  export http_proxy=$HTTP_PROXY
  export https_proxy=$HTTPS_PROXY
  export no_proxy=$NO_PROXY
```

**Comment by kastork** [comment]
Sure.  But if you don't do everything in my GIST, Che simply does not work (it can't pull containers)

**Comment by skabashnyuk** [comment]
@kastork for my understanding. Che is not working on your mac or you just want to get rid of the warning?

**Comment by kastork** [comment]
Last time I checked (back in August) It works if you follow the workarounds in the gist.

 https://gist.github.com/kastork/3d060691e9661064e894c2aa1893e0f4

But those workarounds would not be necessary if Docker for Mac set (or allowed setting) a value for `NO_PROXY`.  Then, as @eivantsov said, Che would pick up the setting and be happy.
